runtime/lua/vim/diagnostic.lua

```diff
@@ -1696,9 +1696,10 @@ M.handlers.virtual_text = {
     end
 
     if opts.virtual_text then
-      if opts.virtual_text.format then
-        diagnostics = reformat_diagnostics(opts.virtual_text.format, diagnostics)
+      local default_format = function(diagnostic)
```

Comment:
The purpose of having a `format` option is exactly for scenarios like this. We won't be adding custom logic to the default code.

Just set this in your config:
```lua
vim.diagnostic.config {
  virtual_text = {
    format = function(d) return d.message:gsub('%s+', ' ') end
  }
}
```

Reply:
I don't understand this to be honest - wouldn't it make sense to provide a *usable* default to all users and make them override the behaviour to something less sensible or personally fitting? I think this is an actual improvement, in all cases.   

Displaying all the whitespace that were intended for editors that don't make the same design decision as neovim in displaying the diagnostics on a single line makes the user experience worse in all cases.

Reply:
> has a funky behavior then the format function is there to massage the input


The server does something totally sensible for multilingual displaying, it uses newlines and indents the next line. 

> Correct me if I'm wrong, but aren't we already removing \n from the text

when I checked, at least the ^@ didn’t show up, so presumably, but maybe the multiple white spaces that stem from the indentation of the next line should still be contracted?

Reply:
I agree with @MariaSolOs , we should avoid adding workarounds to patch up issues in one server. This “solves” the immediate problem with Haskell, but who is to say it doesn’t introduce a problem somewhere else?

Our policy has always been to stick to the specification as written, allowing user customization where possible, and pressure upstream servers to fix their bugs rather than work around them.

> The server does something totally sensible for multiline displaying, it uses newlines and indents the next line.

Nvim can handle multi line diagnostics so why is the Haskell language server collapsing it into a single line? For virtual text we show only the first line of a multiline diagnostic, but all lines are visible in virtual lines or when opening a floating window.

Reply:
If Haskell language server is collapsing lines then that should also be applying this transformation. I assumed Nvim was stripping newlines for virtual text.